lib/nostrum/struct/emoji.ex

```diff
@@ -1,50 +1,170 @@
+  @type user :: User.t | nil
+
+  @typedoc "Whether this emoji must be wrapped in colons"
+  @type require_colons :: boolean | nil
+
+  @typedoc "Whether this emoji is managed"
+  @type managed :: boolean | nil
+
+  @typedoc "Whether this emoji is animated"
+  @type animated :: boolean | nil
 
   @type t :: %__MODULE__{
     id: id,
-    name: name
+    name: name,
+    roles: roles,
+    user: user,
+    require_colons: require_colons,
+    managed: managed,
+    animated: animated
   }
 
-  @derive [Poison.Encoder]
-  defstruct [
-    :id,
-    :name
-  ]
+  @doc ~S"""
+  Formats an emoji struct into an emoji mention.
+
+  Because `Nostrum.Struct.Emoji` implements the `String.Chars` protocol, we 
+  recommend nostrum users to use `String.Chars` instead of this function.
+  See `Nostrum.Struct.Emoji` for more information.
+
+  ## Examples
+
+  ```Elixir
+  emoji = %Nostrum.Struct.Emoji{id: 43819043108, name: "foxbot"}
+
+  Nostrum.Struct.Emoji.format_mention(emoji)
+  "<:foxbot:43819043108>"
+
+  Nostrum.Api.create_message!(4318940318049, "Sending some text with this emoji #{Nostrum.Struct.Emoji.format_mention(emoji)}")
+  ```
+  """
+  @spec format_mention(t) :: String.t
+  def format_mention(emoji)
+  def format_mention(%__MODULE__{id: nil, name: name}), do: name
+  def format_mention(%__MODULE__{animated: true, id: id, name: name}), do: "<a:#{name}:#{id}>"
+  def format_mention(%__MODULE__{id: id, name: name}), do: "<:#{name}:#{id}>"
+
+  @doc ~S"""
+  Formats an emoji struct into its `t:Nostrum.Struct.Emoji.emoji_api_name/0`.
+
+  ## Examples
+
+  ```Elixir
+  emoji = %Nostrum.Struct.Emoji{id: 43819043108, name: "foxbot"}
+
+  api_name = Nostrum.Struct.Emoji.get_api_name(emoji)
+  "foxbot:43819043108"
+
+  Nostrum.Api.create_reaction(471635274857, 47361625345554, api_name)
+  {:ok}
+  ```
+  """
+  @spec get_api_name(t) :: emoji_api_name
+  def get_api_name(emoji)
+  def get_api_name(%__MODULE__{id: nil, name: name}), do: name
+  def get_api_name(%__MODULE__{id: id, name: name}), do: "#{name}:#{id}"
 
   @doc false
   def p_encode do
     %__MODULE__{}
   end
 
-  @doc """
-  Formats a custom emoji for use in a Discord message.
-  """
-  @spec format_custom_emoji(String.t, String.t | integer) :: String.t
-  def format_custom_emoji(name, id) when is_binary(id), do: "<:" <> name <> ":" <> id <> ">"
-  def format_custom_emoji(name, id) do
-    "<:" <> name <> ":" <> to_string(id) <> ">"
-  end
-
   @doc false
-  def to_struct(list) when is_list(list), do: list |> Enum.map(&to_struct(&1))
   def to_struct(map) do
-    struct(__MODULE__, map)
+    new =
+      map
+      |> Map.new(fn {k, v} -> {Util.maybe_to_atom(k), v} end)
+      |> Map.update(:id, nil, &Util.cast(&1, Snowflake))
+      |> Map.update(:roles, nil, &Util.cast(&1, {:list, Snowflake}))
+      |> Map.update(:user, nil, &Util.cast(&1, {:struct, User}))
+
+    struct(__MODULE__, new)
+  end
+end
+
+defimpl String.Chars, for: Nostrum.Struct.Emoji do
```

Comment:
I think we should move this into the module definition.

Reply:
Agreed. Added in the upcoming commit.